proposals/NNNN-member-import-visibility.md

```diff
@@ -0,0 +1,149 @@
+
+## Source compatibility
+
+The proposed change in behavior is source breaking because it adds stricter requirements to name lookup. There is much existing Swift code that will need to be updated to adhere to these new requirements, either by introducing additional import statements in some source files or by reorganizing code among files. This change in behavior therefore must be opt-in, which is why it should be limited to a future language mode with an upcoming feature identifier that allows opt-in with previous language modes.
+
+## ABI compatibility
+
+This change does not affect ABI.
+
+## Implications on adoption
+
+To make it easier to migrate to the new language mode, the compiler can attempt to identify whether a member reference would resolve to a member declared in a transitively imported module and emit a fix-it to suggest adding a direct import to resolve the errors caused by the stricter look up rules:
+
+```swift
+// In this example, RecipeKit is imported in another file
+
+// note: add import of module 'RecipeKit'
+
+let recipe = "1 scoop ice cream, 1 tbs chocolate syrup".parse()
+// error: instance method 'parse()' is inaccessible due to missing import of defining module 'RecipeKit'
+```
+
+With these fix-its, the burden of updating source code to be compatible with the new language mode should be significantly reduced.
+
+This feature will have some impact on source compatibility with older compilers and previous language modes. Adding new direct imports of modules that were previously only transitively imported is a backward compatible change syntactically. However, if the new language mode is necessary in order to make some source code unambiguous, then the ambiguity will become an issue when compiling the same code using an older language mode so maintaining backward compatibility would require additional measures to be taken.
+
+## Future directions
+
+#### Add module qualification syntax for extension members
+
+This proposal seeks to give developers explicit control over which members are visible in a source file because this control can be used to prevent and resolve ambiguities that arise when different modules declare conflicting members in extensions. With this proposal implemented, if an extension member ambiguity still arises in a source file then the developer has the option of curating the imports in that file to resolve the ambiguity. This may work in some situations, but in others it may be awkward to refactor code in order to avoid importing a module that introduces a conflict. For these cases it would be useful to have a syntax that unambiguously identifies the desired extension member at the use site. For example, here's a hypothetical syntax for explicitly calling the `parse()` method declared in the module `RecipeKit`:
+
+```swift
+let recipe = "...".RecipeKit::parse()
+```
```

Comment:
Possibly Rust's turbofish syntax would be a better hypothetical syntax to have here?

Reply:
The syntax I've used here is a [pitched](https://forums.swift.org/t/pitch-fully-qualified-name-syntax/28482) syntax that I also plan to propose at some point. I don't think Rust's turbofish syntax makes sense as an alternative because it is used for specifying generic parameters in Rust. The turbofish syntax would be evocative of generics in Swift too given the similarity to Swift's existing generics syntax.